File: web/src/crypto/asymmetric.ts
/**
 * RSA-4096 asymmetric primitives.
 *
 * Mirrors `src/vellaris/core/asymmetric.py`. Two operations live here:
 *
 * - **OAEP encryption / decryption** for wrapping per-document AES keys
 *   (DEKs) under a recipient's public key. SHA-256 is used both as the
 *   hash and as MGF1's underlying hash. WebCrypto binds MGF1 to the same
 *   hash as the import-time `hash`, so importing with `SHA-256` is enough.
 * - **RSASSA-PSS signing / verification** for the auth challenge. PSS is
 *   a different padding from OAEP — sign/verify must NOT reuse the OAEP
 *   key handles (the original PoC's bug).
 *
 * Keys are exchanged as PEM (PKCS#8 for private, SubjectPublicKeyInfo for
 * public). Both formats round-trip with Python's `cryptography` library.
 */

import { bs } from './_buffer.ts'
import { DecryptError, KeyFormatError, SignatureError } from './errors.ts'

export const KEY_SIZE_BITS = 4096
export const PUBLIC_EXPONENT = new Uint8Array([0x01, 0x00, 0x01]) // 65537

/** PSS salt length in bytes — matches Python's `padding.PSS.DIGEST_LENGTH` for SHA-256. */
export const PSS_SALT_LENGTH = 32

export interface RSAKeyPair {
  readonly privateKey: CryptoKey
  readonly publicKey: CryptoKey
}

/**
 * RSA-4096 keys carry one of two algorithm identifiers depending on usage:
 * `RSA-OAEP` (wrap/unwrap DEKs) or `RSA-PSS` (sign/verify auth challenge).
 * Re-using a key handle across both operations is unsafe (different padding,
 * different attack model). We keep them separate by importing twice when
 * needed, and by generating two distinct keypairs only when the caller asks.
 *
 * For Vellaris we keep ONE underlying RSA private key per user, but wrap it
 * under both algorithm types at use-time via {@link reimportPrivateKeyForPss}.
 */

function pemDecode(pem: string, label: string): Uint8Array {
  const begin = `-----BEGIN ${label}-----`
  const end = `-----END ${label}-----`
  const beginIdx = pem.indexOf(begin)
  const endIdx = pem.indexOf(end)
  if (beginIdx < 0 || endIdx < 0 || endIdx < beginIdx) {
    throw new KeyFormatError(`PEM missing ${label} delimiters`)
  }
  const body = pem.slice(beginIdx + begin.length, endIdx).replace(/\s+/g, '')
  if (!body) {
    throw new KeyFormatError(`PEM body for ${label} is empty`)
  }
  try {
    const binary = atob(body)
    const bytes = new Uint8Array(binary.length)
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i)
    }
    return bytes
  } catch (cause) {
    throw new KeyFormatError(`PEM body for ${label} is not valid base64`, { cause })
  }
}

function pemEncode(der: Uint8Array, label: string): string {
  let binary = ''
  for (let i = 0; i < der.length; i++) {
    binary += String.fromCharCode(der[i])
  }
  const b64 = btoa(binary)
  // Python's `cryptography` writes 64-char lines — match that for byte-equality
  // with fixtures coming out of the CLI / SDK.
  const lines: string[] = []
  for (let i = 0; i < b64.length; i += 64) {
    lines.push(b64.slice(i, i + 64))
  }
  return `-----BEGIN ${label}-----\n${lines.join('\n')}\n-----END ${label}-----\n`
}

/** Generate a fresh RSA-4096 OAEP keypair. The same modulus is used for PSS via re-import. */
export async function generateOaepKeypair(): Promise<RSAKeyPair> {
  const pair = await crypto.subtle.generateKey(
    {
      name: 'RSA-OAEP',
      modulusLength: KEY_SIZE_BITS,
      publicExponent: PUBLIC_EXPONENT,
      hash: 'SHA-256',
    },
    true,
    ['encrypt', 'decrypt'],
  )
  return { privateKey: pair.privateKey, publicKey: pair.publicKey }
}

/** PEM-serialize an RSA private key as unencrypted PKCS#8 (matches Python output). */
export async function serializePrivateKey(privateKey: CryptoKey): Promise<Uint8Array> {
  const der = await crypto.subtle.exportKey('pkcs8', privateKey)
  return new TextEncoder().encode(pemEncode(new Uint8Array(der), 'PRIVATE KEY'))
}

/** PEM-serialize an RSA public key as SubjectPublicKeyInfo (matches Python output). */
export async function serializePublicKey(publicKey: CryptoKey): Promise<Uint8Array> {
  const der = await crypto.subtle.exportKey('spki', publicKey)
  return new TextEncoder().encode(pemEncode(new Uint8Array(der), 'PUBLIC KEY'))
}

/** Load a PEM-encoded RSA private key for OAEP decryption. */
export async function deserializePrivateKeyForOaep(pem: Uint8Array | string): Promise<CryptoKey> {
  const text = typeof pem === 'string' ? pem : new TextDecoder().decode(pem)
  const der = pemDecode(text, 'PRIVATE KEY')
  try {
    return await crypto.subtle.importKey(
      'pkcs8',
      bs(der),
      { name: 'RSA-OAEP', hash: 'SHA-256' },
      true,
      ['decrypt'],
    )
  } catch (cause) {
    throw new KeyFormatError('failed to load private key for OAEP', { cause })
  }
}

/** Load a PEM-encoded RSA private key for PSS signing. Different padding, different key handle. */
export async function deserializePrivateKeyForPss(pem: Uint8Array | string): Promise<CryptoKey> {
  const text = typeof pem === 'string' ? pem : new TextDecoder().decode(pem)
  const der = pemDecode(text, 'PRIVATE KEY')
  try {
    return await crypto.subtle.importKey(
      'pkcs8',
      bs(der),
      { name: 'RSA-PSS', hash: 'SHA-256' },
      true,
      ['sign'],
    )
  } catch (cause) {
    throw new KeyFormatError('failed to load private key for PSS', { cause })
  }
}

/** Load a PEM-encoded RSA public key for OAEP encryption. */
export async function deserializePublicKeyForOaep(pem: Uint8Array | string): Promise<CryptoKey> {
  const text = typeof pem === 'string' ? pem : new TextDecoder().decode(pem)
  const der = pemDecode(text, 'PUBLIC KEY')
  try {
    return await crypto.subtle.importKey(
      'spki',
      bs(der),
      { name: 'RSA-OAEP', hash: 'SHA-256' },
      true,
      ['encrypt'],
    )
  } catch (cause) {
    throw new KeyFormatError('failed to load public key for OAEP', { cause })
  }
}

/** Load a PEM-encoded RSA public key for PSS verification. */
export async function deserializePublicKeyForPss(pem: Uint8Array | string): Promise<CryptoKey> {
  const text = typeof pem === 'string' ? pem : new TextDecoder().decode(pem)
  const der = pemDecode(text, 'PUBLIC KEY')
  try {
    return await crypto.subtle.importKey(
      'spki',
      bs(der),
      { name: 'RSA-PSS', hash: 'SHA-256' },
      true,
      ['verify'],
    )
  } catch (cause) {
    throw new KeyFormatError('failed to load public key for PSS', { cause })
  }
}

/** RSA-OAEP-SHA256 encryption. Used for wrapping AES DEKs per recipient. */
export async function oaepEncrypt(
  plaintext: Uint8Array,
  publicKey: CryptoKey,
): Promise<Uint8Array> {
  const ct = await crypto.subtle.encrypt({ name: 'RSA-OAEP' }, publicKey, bs(plaintext))
  return new Uint8Array(ct)
}

/** RSA-OAEP-SHA256 decryption. */
export async function oaepDecrypt(
  ciphertext: Uint8Array,
  privateKey: CryptoKey,
): Promise<Uint8Array> {
  try {
    const pt = await crypto.subtle.decrypt({ name: 'RSA-OAEP' }, privateKey, bs(ciphertext))
    return new Uint8Array(pt)
  } catch (cause) {
    throw new DecryptError('RSA-OAEP decryption failed', { cause })
  }
}

/** RSASSA-PSS-SHA256 signature with salt_length=32 (matches Python's PSS.DIGEST_LENGTH). */
export async function pssSign(message: Uint8Array, privateKey: CryptoKey): Promise<Uint8Array> {
  const sig = await crypto.subtle.sign(
    { name: 'RSA-PSS', saltLength: PSS_SALT_LENGTH },
    privateKey,
    bs(message),
  )
  return new Uint8Array(sig)
}

/** Verify an RSA-PSS-SHA256 signature. Throws {@link SignatureError} on failure. */
export async function pssVerify(
  message: Uint8Array,
  signature: Uint8Array,
  publicKey: CryptoKey,
): Promise<void> {
  const ok = await crypto.subtle.verify(
    { name: 'RSA-PSS', saltLength: PSS_SALT_LENGTH },
    publicKey,
    bs(signature),
    bs(message),
  )
  if (!ok) {
    throw new SignatureError('RSA-PSS signature verification failed')
  }
}

File: src/vellaris/core/asymmetric.py
"""RSA-4096 asymmetric primitives.

Two operations live here:

- **OAEP encryption / decryption** for wrapping per-document AES keys
  (DEKs) under a recipient's public key. SHA-256 is used both as the
  hash and as MGF1's underlying hash.
- **RSASSA-PSS signing / verification** for the auth challenge. PSS is
  a different padding from OAEP and sign/verify must NOT reuse the
  encryption helpers — the original PoC's bug.

Keys are exchanged as PEM (PKCS#8 for private, SubjectPublicKeyInfo for
public). Other formats are deliberately not supported here.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Final

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding, rsa

from vellaris.core.errors import DecryptError, KeyFormatError, SignatureError

KEY_SIZE_BITS: Final[int] = 4096
"""RSA modulus size in bits."""

PUBLIC_EXPONENT: Final[int] = 65537


@dataclass(frozen=True, slots=True)
class RSAKeyPair:
    """A freshly generated RSA-4096 key pair, in `cryptography` object form."""

    private_key: rsa.RSAPrivateKey
    public_key: rsa.RSAPublicKey


def generate_keypair() -> RSAKeyPair:
    """Generate a fresh RSA-4096 key pair."""
    private_key = rsa.generate_private_key(public_exponent=PUBLIC_EXPONENT, key_size=KEY_SIZE_BITS)
    return RSAKeyPair(private_key=private_key, public_key=private_key.public_key())


def serialize_private_key(private_key: rsa.RSAPrivateKey) -> bytes:
    """Serialize an RSA private key to unencrypted PEM (PKCS#8)."""
    return private_key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    )


def deserialize_private_key(pem: bytes) -> rsa.RSAPrivateKey:
    """Load a PEM-encoded RSA private key. Rejects non-RSA keys."""
    if not isinstance(pem, (bytes, bytearray)):
        raise KeyFormatError(f"private key PEM must be bytes, got {type(pem).__name__}")
    try:
        key = serialization.load_pem_private_key(bytes(pem), password=None)
    except (ValueError, TypeError) as exc:
        raise KeyFormatError(f"failed to load private key: {exc}") from exc
    if not isinstance(key, rsa.RSAPrivateKey):
        raise KeyFormatError(f"expected an RSA private key, got {type(key).__name__}")
    return key


def serialize_public_key(public_key: rsa.RSAPublicKey) -> bytes:
    """Serialize an RSA public key to PEM (SubjectPublicKeyInfo)."""
    return public_key.public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )


def deserialize_public_key(pem: bytes) -> rsa.RSAPublicKey:
    """Load a PEM-encoded RSA public key. Rejects non-RSA keys."""
    if not isinstance(pem, (bytes, bytearray)):
        raise KeyFormatError(f"public key PEM must be bytes, got {type(pem).__name__}")
    try:
        key = serialization.load_pem_public_key(bytes(pem))
    except (ValueError, TypeError) as exc:
        raise KeyFormatError(f"failed to load public key: {exc}") from exc
    if not isinstance(key, rsa.RSAPublicKey):
        raise KeyFormatError(f"expected an RSA public key, got {type(key).__name__}")
    return key


def _oaep_padding() -> padding.OAEP:
    return padding.OAEP(
        mgf=padding.MGF1(algorithm=hashes.SHA256()),
        algorithm=hashes.SHA256(),
        label=None,
    )


def _pss_padding() -> padding.PSS:
    # Salt length matches the hash output, the conventional choice.
    return padding.PSS(
        mgf=padding.MGF1(algorithm=hashes.SHA256()),
        salt_length=padding.PSS.DIGEST_LENGTH,
    )


def oaep_encrypt(plaintext: bytes, public_key: rsa.RSAPublicKey) -> bytes:
    """RSA-OAEP-SHA256 encryption. Used for wrapping AES DEKs per recipient."""
    return public_key.encrypt(plaintext, _oaep_padding())


def oaep_decrypt(ciphertext: bytes, private_key: rsa.RSAPrivateKey) -> bytes:
    """RSA-OAEP-SHA256 decryption."""
    try:
        return private_key.decrypt(ciphertext, _oaep_padding())
    except ValueError as exc:
        raise DecryptError(f"RSA-OAEP decryption failed: {exc}") from exc


def pss_sign(message: bytes, private_key: rsa.RSAPrivateKey) -> bytes:
    """RSASSA-PSS-SHA256 signature. Used for the auth challenge."""
    return private_key.sign(message, _pss_padding(), hashes.SHA256())


def pss_verify(message: bytes, signature: bytes, public_key: rsa.RSAPublicKey) -> None:
    """Verify an RSASSA-PSS-SHA256 signature. Raises :class:`SignatureError` on failure."""
    try:
        public_key.verify(signature, message, _pss_padding(), hashes.SHA256())
    except InvalidSignature as exc:
        raise SignatureError("RSA-PSS signature verification failed") from exc
